src/blocks.rs

```diff
@@ -0,0 +1,349 @@
+use std::ops::Deref;
+use std::collections::{BTreeSet, BTreeMap, HashMap};
+use std::borrow::Borrow;
+
+use block::{BlockId, Block, Vote, is_quorum_of};
+use name::{Name, Prefix};
+
+pub type ValidBlocks = BTreeSet<BlockId>;
+pub type CurrentBlocks = BTreeSet<BlockId>;
+
+/// Mapping from votes to voters: (vote.from -> (vote.to -> names)).
+pub type VoteCounts = BTreeMap<BlockId, BTreeMap<BlockId, BTreeSet<Name>>>;
+
+pub struct Blocks(HashMap<BlockId, Block>);
+
+impl Deref for Blocks {
+    type Target = HashMap<BlockId, Block>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Blocks {
+    pub fn new() -> Blocks {
+        Blocks(HashMap::new())
+    }
+
+    pub fn insert(&mut self, block: Block) -> BlockId {
+        let id = block.get_id();
+        self.0.insert(id, block);
```

Comment:
I wonder how robust the `DefaultHasher` is to collisions, we should maybe check that any existing value in the map is `==` to `Block`.